### PR TITLE
samcli: add local autoinstall for linux/windows

### PR DESCRIPTION
**Pull request by gsweetwood**
## Problem
 Many users don't have SAM CLI and/or don't want to follow the several steps required to install it.

## Solution
If the SAM CLI is not found, install it locally. Currently scoped to just Linux users.

## License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/shared/utilities/cliUtils.ts [hunk 1]
```diff
@@ -314,6 +335,53 @@ async function installSsmCli(
...
+        }
+        const dirname = path.dirname(samInstaller)
+        new admZip(samInstaller).extractAllTo(dirname, true)
+
+        const result = await new ChildProcess('sh', [
+            path.join(dirname, 'install'),
+            '-u',
+            '-i',
+            outDir,
+            '-b',
```
Comment (justinmk3) [inline src/shared/utilities/cliUtils.ts:372]: short args like `-i` and `-b` are convenience aliaes for interactive usage. For non-interactive usage, it's usually helpful to use the long names (except perhaps for some very well-known commands).

File: src/test/shared/sam/cli/samCliConfiguration.test.ts [hunk 2]
```diff
@@ -74,6 +75,9 @@ describe('samCliConfiguration', function () {
             settingsConfiguration
         )
 
+        if (os.platform() === 'linux') {
+            this.skip()
```
Comment (justinmk3) [inline src/test/shared/sam/cli/samCliConfiguration.test.ts:79]: this test worked before this PR. why is it skipped always on linux now?

**Review by justinmk3 — COMMENTED** [review]
Next step is to look for a way to get this working on Windows, perhaps by installing to a higher-level directory (e.g. ~/.aws/toolkit/)

**Comment by gsweetwood** [comment]
Added a small refactor to the `samCliLocator` to search locally if not found globally. (reversed for C9)

File: src/test/shared/sam/cli/samCliConfiguration.test.ts (hunk 2 above)
Comment (gsweetwood) [reply src/test/shared/sam/cli/samCliConfiguration.test.ts]: Now that I added the option to install or not if the CLI is not detected, we can run this test on all platforms. This test was assuring no cli path was found, but it was autoinstalling on Linux, so it failed.

**Comment by gsweetwood** [comment]
Added the installation logic for Windows. Needed to use `~/.aws/toolkits` to avoid the path length error. I did try the shorter `/sam` instead of `AWSSAMCLI/bin/sam` but it wasn't enough. Symlinks DO NOT work for executables in Windows, so the cli path needs to be the same location.

**Pull request closed without merging**